platforms/software/dependency-management/src/main/java/org/gradle/api/internal/artifacts/ivyservice/resolveengine/graph/builder/NodeState.java

```diff
@@ -229,92 +252,99 @@ public boolean isTransitive() {
+        ) {
+            // Our excludes did not change, or after applying new excludes to our outgoing dependencies,
+            // the filtered dependencies did not change. We have the same dependencies as the previous traversal.
 
-        if (previousTraversalExclusions == null) {
-            // We have not visited this node's dependencies yet. Visit all outgoing dependencies.
-            visitAllDependencies(resolutionFilter, discoveredEdges);
-        } else if (virtualPlatformNeedsRefresh) {
-            // Virtual platforms require their constraints to be recomputed each time, as each module addition
-            // can cause a shift in versions. Therefore, we perform a full dependency visit even though
-            // we've already visited this node's dependencies before.
-            clearDependencyVisitState();
-            visitAllDependencies(resolutionFilter, discoveredEdges);
-        } else if (previousTraversalExclusions.equals(resolutionFilter) || (doesNotHaveDependencies && !dependenciesMayChange)) {
-            // The excludes did not change, or we have no dependencies at all. Skip normal visiting.
-            visitNewAndInvalidatedDependencies(resolutionFilter, discoveredEdges);
-        } else if (cachedFilteredDependencyStates != null && !recomputeOutgoingDependencies(resolutionFilter)) {
-            // The excludes changed, and after applying the new excludes to our outgoing dependencies, the
-            // filtered dependencies did not change. We can skip normal dependency visit logic and just update the edges
-            // that need to be updated. However, we will also be skipping other logic that updates the resolution filter
-            // on our outgoing edges. So, do that here instead.
-            for (EdgeState outgoingEdge : outgoingEdges) {
-                outgoingEdge.updateTransitiveExcludesAndRequeueTargetNodes(resolutionFilter);
+            if (!sameExcludes) {
+                // Our excludes changed. Update our outgoing edges with the new excludes.
+                for (EdgeState outgoingEdge : outgoingEdges) {
+                    outgoingEdge.updateTransitiveExcludesAndRequeueTargetNodes(resolutionFilter);
+                }
+                if (virtualEdges != null) {
+                    for (EdgeState virtualEdge : virtualEdges) {
+                        virtualEdge.updateTransitiveExcludesAndRequeueTargetNodes(resolutionFilter);
+                    }
+                }
             }
 
-            visitNewAndInvalidatedDependencies(resolutionFilter, discoveredEdges);
-        } else {
-            // Our transitive excludes changed, which changed which direct dependencies we are targeting.
-            // Clear any old visit state and perform a full visit of our dependencies.
-            clearDependencyVisitState();
-            visitAllDependencies(resolutionFilter, discoveredEdges);
+            if (!ancestorsStrictVersions.equals(previousAncestorsStrictVersions)) {
```

Comment:
🎉 That was one of the points I had in mind for later discussion: propagate down the graph.